### PR TITLE
Teamraum visible users

### PR DESCRIPTION
**Pull request by elioschmutz**
This PR implements a special protection for users, groups, vocabularies, sources and endpoints within a teamraum.

This is required because a teamraum has a complete different security policy than a gever has. 

For [CA-3592]

## Checklist

_Everything has to be done/checked. Checked but not present means the author deemed it unnecessary._

- |x] Changelog entry
- [x] Link to issue (Jira or GitHub) and backlink in issue (Jira)

_Only applicable should be left and checked._

- Upgrade-Steps:
  - [ ] SQL Operations do not use imported model (see [docs](https://4teamwork.atlassian.net/wiki/spaces/4TEAM/pages/994344994/Upgrade-Steps))
  - [ ] Make it deferrable if possible
  - [ ] Execute as much as possible conditionally
  - DB-Schema migration
    - [ ] All changes on a model (columns, etc) are included in a DB-schema migration.
    - [ ] Constraint names are shorter than 30 characters (`Oracle`)
- API change:
  - [ ] Documentation is updated
  - [ ] API Changelog entry (see [guide](https://4teamwork.atlassian.net/wiki/spaces/4TEAM/pages/451248812/API+Changelog+Guidelines))
  - If breaking:
    - [ ] api-change label added
    - [ ] #delivery channel notified about breaking change
    - [ ] Scrum master is informed
- Bug fixed:
  - [ ] Resolved any Sentry issues caused by this bug
- New functionality:
  - [ ] for `document` also works for `mail`
  - [ ] for `task` also works for `forwarding`
- Further improvements needed:
  - [ ] Create follow-up stories and link them in the PR and Jira issue
- [ ] Change could impact client installations, client policies need to be adapted
- New translations
  - [ ] All msg-strings are unicode
- Change in schema definition:
  - [ ] If `missing_value` is specified, then `default` has to be set to the same value

[CA-3592]: https://4teamwork.atlassian.net/browse/CA-3592?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: opengever/workspace/__init__.py [hunk 1]
```diff
@@ -7,6 +7,17 @@
 
 WHITELISTED_TEAMRAUM_GLOBAL_SOURCES = []
 
+WHITELISTED_TEAMRAUM_PORTAL_TYPES = [
```
Comment (buchi) [inline opengever/workspace/__init__.py:10]: make this a `set` for a more efficient containment check

File: opengever/base/visible_users_and_groups_filter.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
+from opengever.ogds.models.service import ogds_service
+from opengever.workspace import is_workspace_feature_enabled
+from plone import api
+
+
+class VisibleUsersAndGroupsFilter:
+    """This singleton object provides functions to check if a
```
Comment (buchi) [inline opengever/base/visible_users_and_groups_filter.py:7]:
```suggestion
    """Provides functions to check if a
```

The singleton part is not relevant here nor does this class implement the singleton pattern.

File: opengever/base/visible_users_and_groups_filter.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    specific user or group should be visible to the currently logged in user
+    or not.
+
+    In general, every user can get information about every other user and group
```
Comment (buchi) [inline opengever/base/visible_users_and_groups_filter.py:11]: I would try to shorten the rest of the docstring, e.g.:

"In GEVER user and group information is exposed to all users independant of their access rights.
In teamraum users are only allowed to get user information about users that participate in workspaces the user has access to."

**Review by buchi — COMMENTED** [review]
LGTM, just some minor suggestions, mainly wording...
I still have to do a performance check.

**Comment by elioschmutz** [comment]
@buchi 👍  I updated all of your change requests.

**Comment by buchi** [comment]
@elioschmutz I've done some basic performance checks. It's not as bad as expected because in teamraum there are not many calls to the protected endpoints.

I've noticed that adding todo gets noticeably slower and `can_access_principal()` is called multiple times. Thus I would suggest to cache `get_whitelisted_principals()` on the request, which is quite simple und should be sufficient.

Unfortunately the `@view.memoize_contextless` decorator can't be used as the VUAGF doesn't have a `request` attribute (could be added though). But you could do something like this:
```
def get_whitelisted_principals(self):
    request = getRequest()
    allowed_users_and_groups = getattr(request, '_vuagf_allowed_users_and_groups', None)
    if allowed_users_and_groups is None:
        allowed_users_and_groups = set(self.extract_allowed_users_and_groups())
        allowed_users_and_groups |= set(self.extract_group_members(allowed_users_and_groups))
        allowed_users_and_groups.add(api.user.get_current().getId())
        request._vuagf_allowed_users_and_groups = allowed_users_and_groups
    return allowed_users_and_groups
```

**Comment by buchi** [comment]
An acronym for VisibleUsersAndGroupsFilter 😉

**Comment by elioschmutz** [comment]
👍 @buchi I added caching. See my last commit

**Review by buchi — APPROVED** [review]

**Pull request merged**